btc_light_client.go

```diff
@@ -46,25 +48,60 @@ func (lc *BTCLightClient) FindPreviousCheckpoint() (blockchain.HeaderCtx, error)
 	return nil, nil
 }
 
-func (lc *BTCLightClient) AddHeader(height int64, header wire.BlockHeader) error {
-	return lc.btcStore.AddHeader(height, header)
+func (lc *BTCLightClient) SetHeader(height int64, header wire.BlockHeader) error {
+	return lc.btcStore.SetHeader(height, header)
 }
 
 // We assume we always insert valid header. Acctually, Cosmos can revert a state
 // when module return error so this assumtion is reasonable
 func (lc *BTCLightClient) InsertHeaders(headers []wire.BlockHeader) error {
 	latestHeight := lc.btcStore.LatestHeight()
-	for _, header := range headers {
+
+	return lc.insertHeaderStartAtHeight(uint64(latestHeight), headers)
+}
+
+
+func (lc *BTCLightClient) insertHeaderStartAtHeight(startHeight uint64, headers []wire.BlockHeader) error {
+	height := startHeight + 1
+	for i, header := range headers {
 		if err := lc.CheckHeader(header); err != nil {
-			return err
+			return NewInvalidHeaderErr(header.BlockHash().String(), i)
 		}
 
-		latestHeight = latestHeight + 1
-		lc.AddHeader(latestHeight, header)
+		// override a height
+		lc.SetHeader(int64(height), header)
+		height = height + 1
 	}
 	return nil
 }
 
+func (lc *BTCLightClient) computeTotalWorkFork(startBlock *LightBlock, headers []wire.BlockHeader) *big.Int {
+	totalWork := startBlock.TotalWork
+	for _, header := range headers {
+		totalWork.Add(totalWork, blockchain.CalcWork(header.Bits))
+	}
+	return totalWork
+}
+
+func (lc *BTCLightClient) HandleFork(headers []wire.BlockHeader) error {
+	// find the light block match with first header
+	firstHeader := headers[0]
+	if lightBlock := lc.btcStore.LightBlockByHash(firstHeader.BlockHash()); lightBlock != nil {
+		currentTotalWork := lc.btcStore.LatestLightBlock()
+		otherForkTotalWork := lc.computeTotalWorkFork(lightBlock, headers[1:])
+
+		// comapre with current fork
+		if otherForkTotalWork.Cmp(currentTotalWork.TotalWork) > 0 {
+			// accepte new fork
+			return lc.insertHeaderStartAtHeight(uint64(lightBlock.Height), headers[1:])
+		} else {
+			return errors.New("Invalid fork")
+		}
+	}
+	return errors.New("Header doesn't belong to the chain")
```

Comment:
In Go, we prefer to do early exit, rather than packing the whole logic inside `if`

```suggestion
	lightBlock := lc.btcStore.LightBlockByHash(firstHeader.BlockHash())
	if lightBlock == nil {
		return errors.New("Header doesn't belong to the chain")
	}
		currentTotalWork := lc.btcStore.LatestLightBlock()
		otherForkTotalWork := lc.computeTotalWorkFork(lightBlock, headers[1:])

		// comapre with current fork
		if otherForkTotalWork.Cmp(currentTotalWork.TotalWork) > 0 {
			return lc.insertHeaderStartAtHeight(uint64(lightBlock.Height), headers[1:])
		} else {
			return errors.New("Invalid fork")
		}
```

Reply:
Noted!